lib/internal/modules/esm/hooks.js

```diff
@@ -110,6 +112,23 @@ class Hooks {
   // Cache URLs we've already validated to avoid repeated validation
   #validatedUrls = new SafeSet();
 
+  /**
+   * Import and register custom/user-defined module loader hook(s).
+   * @param {string} urlOrSpecifier
+   */
+  async register(urlOrSpecifier) {
+    const moduleLoader = require('internal/process/esm_loader').esmLoader;
+
+    const parentURL = pathToFileURL(getCwdSafe()).href;
```

Comment:
That's the culprit, we shouldn't use the cwd as parent URL

Reply:
After reading the code a bit more, I'm not sure where you'd get that URL if not given by the user. Using the cwd is a quite surprising thing I think, folks would not expect that. It's especially weird when specifying a package name, as it's going to be resolved very differently depending on where the script is run from.

```sh
mkdir repro
cd repro
mkdir -p src/node_modules/loader
mkdir -p lib/node_modules/loader
echo 'console.log("loader v1.0")' > src/node_modules/loader/index.js
echo 'console.log("loader v2.0")' > lib/node_modules/loader/index.js
echo 'require("node:module").register("loader")' > src/index.js
(cd src && node --loader 'data:text/javascript,' ../src/index.js) # prints "loader v1.0"
(cd lib && node --loader 'data:text/javascript,' ../src/index.js) # prints "loader v2.0"
cd ..
rm -r repro
```

I suggest we add another parameter to the `register` method to let folks provide the parentURL they want the loader specifier resolved from. Alternatively we could restrict to absolute URLs until we find a better solution.

Reply:
No, `path.resolve` uses the cwd, and also paths instead of URLs.